### PR TITLE
Add plugin_paths parameter to terraform module.

### PR DESCRIPTION
**Pull request by teamosceola**
##### SUMMARY
The list `plugin_paths` is used in the `init` phase of Terraform by setting the `-plugin-dir` command line argument to a path that contains Terraform plugins.
When the `-plugin-dir` argument is set, the plugin discovery and auto-download of Terraform is disabled.
This is useful when running Terraform in automation environments for testing where the production endpoints are strictly controlled or on air-gaped networks and you need to control the versions of plugins available.

##### ISSUE TYPE
- Feature Pull Request

##### COMPONENT NAME
terraform

##### ADDITIONAL INFORMATION
The directory structure in the plugin path can be tricky. The [Terraform docs](https://learn.hashicorp.com/tutorials/terraform/automate-terraform#pre-installed-plugins) show a simple directory of files, but in testing that doesn't work, the directory structure has to follow the same structure you would see if Terraform auto-downloaded the plugins.

For example, if you used `project_paths: [ '/path/to/providers' ]`, the directory structure for `/path/to/providers` would need to look like this: (`tree /path/to/providers`)
```
/path/to/providers/
└── registry.terraform.io
    └── hashicorp
        └── vsphere
            ├── 1.24.0
            │   └── linux_amd64
            │       └── terraform-provider-vsphere_v1.24.0_x4
            └── 1.26.0
                └── linux_amd64
                    └── terraform-provider-vsphere_v1.26.0_x4
```

### CODE REVIEW

**Comment by m-yosefpor** [comment]
Thanks for your contribution!
Please add a [changelog fragment](https://docs.ansible.com/ansible/latest/community/development_process.html#creating-a-changelog-fragment).

**Comment by teamosceola** [comment]
> Thanks for your contribution!
> Please add a [changelog fragment](https://docs.ansible.com/ansible/latest/community/development_process.html#creating-a-changelog-fragment).

Done!

File: changelogs/fragments/2308-terraform-add-plugin_paths-parameter.yaml [hunk 1]
```diff
@@ -0,0 +1,3 @@
+---
+minor_changes:
+  - terraform - add `plugin_paths` parameter which allows disabling Terraform from performing plugin discovery and auto-download (https://github.com/ansible-collections/community.general/pull/2308).
```
Comment (felixfontein) [inline changelogs/fragments/2308-terraform-add-plugin_paths-parameter.yaml:3]:
```suggestion
  - terraform - add ``plugin_paths`` parameter which allows disabling Terraform from performing plugin discovery and auto-download (https://github.com/ansible-collections/community.general/pull/2308).
```

File: plugins/modules/cloud/misc/terraform.py [hunk 2]
```diff
@@ -33,6 +33,14 @@
...
     type: path
     required: true
+  plugin_paths:
+    description:
+      - List of paths containing Terraform plugin executable files.
+      - Plugin executables can be downloaded from https://releases.hashicorp.com/
+      - When set, the plugin discovery and auto-download behavior of Terraform is disabled.
+    type: list
+    elements: path
+    required: false
```
Comment (felixfontein) [inline plugins/modules/cloud/misc/terraform.py:43]:
```suggestion
    version_added: 3.0.0
```

File: plugins/modules/cloud/misc/terraform.py [hunk 3]
```diff
@@ -33,6 +33,14 @@
         vars.tf/main.tf/etc to use.
     type: path
     required: true
+  plugin_paths:
+    description:
+      - List of paths containing Terraform plugin executable files.
+      - Plugin executables can be downloaded from https://releases.hashicorp.com/
```
Comment (felixfontein) [inline plugins/modules/cloud/misc/terraform.py:39]:
```suggestion
      - Plugin executables can be downloaded from U(https://releases.hashicorp.com/).
```

**Comment by felixfontein** [comment]
> The directory structure in the plugin path can be tricky. The [Terraform docs](https://learn.hashicorp.com/tutorials/terraform/automate-terraform#pre-installed-plugins) show a simple directory of files, but in testing that doesn't work, the directory structure has to follow the same structure you would see if Terraform auto-downloaded the plugins.
> 
> For example, if you used `project_paths: [ '/path/to/providers' ]`, the directory structure for `/path/to/providers` would need to look like this: (`tree /path/to/providers`)
> 
> ```
> /path/to/providers/
> └── registry.terraform.io
>     └── hashicorp
>         └── vsphere
>             ├── 1.24.0
>             │   └── linux_amd64
>             │       └── terraform-provider-vsphere_v1.24.0_x4
>             └── 1.26.0
>                 └── linux_amd64
>                     └── terraform-provider-vsphere_v1.26.0_x4
> ```

Shouldn't this information be included in the documentation as well? (Maybe in the example, because the tree formatting will survive there?)

File: plugins/modules/cloud/misc/terraform.py [hunk 4]
```diff
@@ -295,6 +315,7 @@ def main():
         argument_spec=dict(
             project_path=dict(required=True, type='path'),
             binary_path=dict(type='path'),
+            plugin_paths=dict(required=False, type='list', elements='path', default=None),
```
Comment (aminvakil) [inline plugins/modules/cloud/misc/terraform.py:318]:
```suggestion
            plugin_paths=dict(type='list', elements='path'),
```

**Review by aminvakil — COMMENTED** [review]
Thanks for adding this!

**Comment by teamosceola** [comment]
> > The directory structure in the plugin path can be tricky. The [Terraform docs](https://learn.hashicorp.com/tutorials/terraform/automate-terraform#pre-installed-plugins) show a simple directory of files, but in testing that doesn't work, the directory structure has to follow the same structure you would see if Terraform auto-downloaded the plugins.
> > For example, if you used `project_paths: [ '/path/to/providers' ]`, the directory structure for `/path/to/providers` would need to look like this: (`tree /path/to/providers`)
> > ```
> > /path/to/providers/
> > └── registry.terraform.io
> >     └── hashicorp
> >         └── vsphere
> >             ├── 1.24.0
> >             │   └── linux_amd64
> >             │       └── terraform-provider-vsphere_v1.24.0_x4
> >             └── 1.26.0
> >                 └── linux_amd64
> >                     └── terraform-provider-vsphere_v1.26.0_x4
> > ```
> 
> Shouldn't this information be included in the documentation as well? (Maybe in the example, because the tree formatting will survive there?)

What about something like this?
```
EXAMPLES = """
- name: Basic deploy of a service
  community.general.terraform:
    project_path: '{{ project_dir }}'
    state: present

- name: Define the backend configuration at init
  community.general.terraform:
    project_path: 'project/'
    state: "{{ state }}"
    force_init: true
    backend_config:
      region: "eu-west-1"
      bucket: "some-bucket"
      key: "random.tfstate"

- name: Define the backend configuration with one or more files at init
  community.general.terraform:
    project_path: 'project/'
    state: "{{ state }}"
    force_init: true
    backend_config_files:
      - /path/to/backend_config_file_1
      - /path/to/backend_config_file_2

- name: Disable plugin discovery and auto-download by setting plugin_paths
  community.general.terraform:
    project_path: 'project/'
    state: "{{ state }}"
    force_init: true
    plugin_paths:
      - /path/to/plugins_dir_1
      - /path/to/plugins_dir_2

### Example directory structure for plugin_paths example
# $ tree /path/to/plugins_dir_1
# /path/to/plugins_dir_1/
# └── registry.terraform.io
#     └── hashicorp
#         └── vsphere
#             ├── 1.24.0
#             │   └── linux_amd64
#             │       └── terraform-provider-vsphere_v1.24.0_x4
#             └── 1.26.0
#                 └── linux_amd64
#                     └── terraform-provider-vsphere_v1.26.0_x4
"""
```

File: plugins/modules/cloud/misc/terraform.py [hunk 5]
```diff
@@ -33,6 +33,19 @@
...
     type: path
     required: true
+  plugin_paths:
+    description:
+      - List of paths containing Terraform plugin executable files.
+      - Plugin executables can be downloaded from U(https://releases.hashicorp.com/).
+      - When set, the plugin discovery and auto-download behavior of Terraform is disabled.
+      - The directory structure in the plugin path can be tricky. The Terraform docs
+        U(https://learn.hashicorp.com/tutorials/terraform/automate-terraform#pre-installed-plugins)
+        show a simple directory of files, but in testing that doesn't work, the directory structure
```
Comment (felixfontein) [inline plugins/modules/cloud/misc/terraform.py:43]:
```suggestion
        show a simple directory of files, but in testing that does not work, the directory structure
```

**Review by felixfontein — COMMENTED** [review]
The example looks great!

**Comment by felixfontein** [comment]
@m-yosefpor @aminvakil do you have further comments, or should we get this merged?

File: plugins/modules/cloud/misc/terraform.py [hunk 6]
```diff
@@ -33,6 +33,19 @@
...
     type: path
     required: true
+  plugin_paths:
+    description:
+      - List of paths containing Terraform plugin executable files.
+      - Plugin executables can be downloaded from U(https://releases.hashicorp.com/).
+      - When set, the plugin discovery and auto-download behavior of Terraform is disabled.
+      - The directory structure in the plugin path can be tricky. The Terraform docs
+        U(https://learn.hashicorp.com/tutorials/terraform/automate-terraform#pre-installed-plugins)
+        show a simple directory of files, but in testing that does not work, the directory structure
```
Comment (aminvakil) [inline plugins/modules/cloud/misc/terraform.py:43]: I'm not sure if "in testing that does not work" is a good sentence here, maybe "in action" or "actually"?

**Review by aminvakil — COMMENTED** [review]
Thanks for adding a complete description!

File: plugins/modules/cloud/misc/terraform.py (hunk 6 above)
Comment (teamosceola) [reply plugins/modules/cloud/misc/terraform.py]: Went with "actually"

**Review by aminvakil — APPROVED** [review]
LGTM

**Comment by russoz** [comment]
LGTM

**Comment by felixfontein** [comment]
@teamosceola thanks for implementing this!
@m-yosefpor @aminvakil @russoz thanks for reviewing!

**Comment by Andersson007** [comment]
@felixfontein there are no labels. Should it be backported?

**Comment by aminvakil** [comment]
@Andersson007 This is a new feature, why do you think it should be backported?

**Comment by Andersson007** [comment]
@aminvakil maybe something has changed but in the past the collection was released from a special branch, not from main. Let's see what felixfontein will say

**Comment by felixfontein** [comment]
@Andersson007 the last 2.x.0 minor release is out, so no more backporting of features until `stable-3` exists (which will happen on Monday, when 3.0.0 is released).

**Comment by felixfontein** [comment]
(The 'no backporting of features' always happens between the last X.y.0 minor release and the (X+1).0.0 major release, which is something that doesn't happen very often though :) )

**Comment by Andersson007** [comment]
@felixfontein ah, I see now, thanks for clarifying

**Pull request merged**